torch/fx/passes/runtime_assert.py

```diff
@@ -85,60 +93,140 @@ def insert_deferred_runtime_asserts(
+                torch.ops.aten.sym_stride.int,
+                torch.ops.aten.storage_offset.default,
+            )
+            and node.args[0].op != "placeholder"  # type: ignore[union-attr]
+        ):  # export: sym_size.int(node, dim)
+            val = _get_sym_val(node)
+        if (
+            node.target == operator.getitem
+            and node.args[0].target in (
+                "size",
+                "stride",
+                "storage_offset",
+            )
+            and node.args[0].args[0].op != "placeholder"
+        ):  # dynamo export: getitem(size(node), dim)
+            val = _get_sym_val(node)
+        return (
+            val is not None
+            and not isinstance(val, sympy.Number)
+        )
 
     # Identify what symbols we need to reify.  This isn't strictly needed
     # but helps reduce churn on the graph
     needed_symbols: Set[sympy.Symbol] = set()
     for ras in ras_by_symbol.values():
         for ra in ras:
             needed_symbols.update(free_symbols(ra.expr))
+    for node in gm.graph.nodes:
+        if (
+            node.op == "call_function"
+            and _is_intermediate_tensor_shape_call(node)
+        ):  # use symbols for intermediate shape calls, even if not in runtime asserts
+            sym_expr = _get_sym_val(node)
+            needed_symbols.update(sym_expr.free_symbols)
 
     log.debug("needed_symbols = %s", needed_symbols)
 
+    # Track asserts/checks we've added
+    added_asserts: Set["sympy.Expr"] = set()
+    constrained_unbacked_symbols: Set["sympy.Symbol"] = set()
+
+    def _sympy_interp(symbol_to_proxy, expr):
+        # sympy_interp() with hash consing
+        from sympy import Integer, Number, Symbol
+        from sympy.logic.boolalg import BooleanAtom
+        from torch.utils._sympy.interp import _run_sympy_handler, sympy_interp
+
+        # base cases, don't cache
+        if isinstance(expr, (Integer, Number, Symbol, BooleanAtom)):
+            return sympy_interp(
+                PythonReferenceAnalysis, symbol_to_proxy, expr
+            )
+        # hash cons
+        if expr in symbol_to_proxy:
+            return symbol_to_proxy[expr]
+
+        # hash cons on arguments
+        fx_args = [
+            (
+                symbol_to_proxy[arg]
+                if arg in symbol_to_proxy
+                else _sympy_interp(symbol_to_proxy, arg)
+            )
+            for arg in expr.args
+        ]
+
+        # run expr handler
+        symbol_to_proxy[expr] = _run_sympy_handler(
+            PythonReferenceAnalysis, fx_args, expr
+        )
+        return symbol_to_proxy[expr]
+
+    def _is_bound_expr_for_symbol(expr: "sympy.Expr") -> bool:
+        # This is probably unnecessary, but since torch._check() calls for single-symbol bounds
+        # like u0 >= 0, 10 >= u0 accumulate range info in the ShapeEnv, and we insert sym_constrain_range calls
+        # anyways, we designate these calls as redundant and remove them.
```

Comment:
I... sort of see what you're going for here? But I'm not exactly sure how you're doing it / maybe there is a better way.